yt_dlp/extractor/bbc.py

```diff
@@ -1235,26 +1359,92 @@ def parse_media(media):
                         'timestamp': item_time,
                         'description': strip_or_none(item_desc),
                     })
+
             for resp in (initial_data.get('data') or {}).values():
                 name = resp.get('name')
                 if name == 'media-experience':
                     parse_media(try_get(resp, lambda x: x['data']['initialItem']['mediaItem'], dict))
                 elif name == 'article':
-                    for block in (try_get(resp,
-                                          (lambda x: x['data']['blocks'],
-                                           lambda x: x['data']['content']['model']['blocks'],),
-                                          list) or []):
-                        if block.get('type') not in ['media', 'video']:
-                            continue
-                        parse_media(block.get('model'))
+                    for block in traverse_obj(resp, (
+                            'data', (None, ('content', 'model')), 'blocks',
+                            lambda _, v: v.get('type') in {'media', 'video'},
+                            'model', {dict})):
+                        parse_media(block)
             return self.playlist_result(
                 entries, playlist_id, playlist_title, playlist_description)
 
+        # extract from SIMORGH_DATA hydration JSON
+        simorgh_data = self._search_json(
+            r'window\s*\.\s*SIMORGH_DATA\s*=', webpage,
+            'simorgh data', playlist_id, default={})
+        if simorgh_data:
+            done = False
+            for video_data in traverse_obj(simorgh_data, (
+                    'pageData', 'content', 'model', 'blocks', is_type('video', 'legacyMedia'))):
+                model = traverse_obj(video_data, (
+                    'model', 'blocks', is_type('aresMedia'),
+                    'model', 'blocks', is_type('aresMediaMetadata'),
+                    'model', {dict}, any))
+                if video_data['type'] == 'video':
+                    entry = parse_model(model)
+                else:  # legacyMedia: no duration, subtitles
+                    block_id, entry = traverse_obj(model, ('blockId', {str})), None
+                    media_data = traverse_obj(simorgh_data, (
+                        'pageData', 'promo', 'media',
+                        {lambda x: x if x['id'] == block_id else None}))
+                    formats = traverse_obj(media_data, ('playlist', lambda _, v: v['url'], {
+                        'url': ('url', {url_or_none}),
+                        'ext': ('format', {str}),
+                        'tbr': ('bitrate', {k_int_or_none}),
+                    }, {lambda u: u.get('url') and u}))
```

Comment:
```suggestion
                    }))
```
why?

Reply:
The traversal result must only include dicts with a valid ['url'] (that previously passed url_or_none).

Reply:
ok, leaving as is, then.

Reply:
So this?
```diff
-                    formats = traverse_obj(media_data, ('playlist', lambda _, v: v['url'], {
+                    formats = traverse_obj(media_data, ('playlist', lambda _, v: url_or_none(v['url']), {
                         'url': ('url', {url_or_none}),
                         'ext': ('format', {str}),
                         'tbr': ('bitrate', {k_int_or_none}),
-	                 }, {lambda u: u.get('url') and u}))
+                    })
```

Reply:
Should there be a `utils.has_value_or_none()` helper, s/t like this?
```py
def has_value_or_none(*keys):
    return lambda x: x if all(x.get(k) for k in keys) else None
```

Reply:
Why? Pls explain. I was just saying how this is unnecessary here...

Reply:
I would suggest
```diff
-                    formats = traverse_obj(media_data, ('playlist', lambda _, v: v['url'], {
-                        'url': ('url', {url_or_none}),
+                    formats = traverse_obj(media_data, ('playlist', lambda _, v: url_or_none(v['url']), {
+                        'url': 'url',
                         'ext': ('format', {str}),
                         'tbr': ('bitrate', {k_int_or_none}),
-	                 }, {lambda u: u.get('url') and u}))
+                    })
```
since we already tested for `url_or_none`

Reply:
Does that prevent formats with no `url` key?

Reply:
`lambda _, v: url_or_none(v['url'])` is a filter function, filtering for a truthy value. `url_or_none(v['url'])` will be truthy if the dict contains a url at the `'url'` key.

`url_or_none` currently `str.strip`s values. So this might still be relevant, like dirk said. I was under the impression `url_or_none` should not change the value and only be a filter (and in fact helpers should behave this way imo). That means that my comment is void though

Reply:
Eg, `{int}` filters for type, but `{int_or_none}` makes a numeric `str` into an `int`.

Generally, a lot of traversals are filtering with `{str}`. Are there cases where that might remove an underlying JSON number that ought to be stringified (as in upstream `txt_or_none()`)?

Reply:
Imo its better to filter than to convert, snice everything converts to `str`. This means that
```json
{"description": {"short": "short description", "full": "longer description"}}
```

will not result in a dict getting stringified.

Reply:
The application for key `{type}` is `x if isinstance(x, type) else None`, not `type(x)`, no?

Reply:
yes. Whats the problem with that? i was saying we DONT want to stringify dicts like above?

Reply:
>Generally, a lot of traversals are filtering with {str}. ...

So (not necessarily relevant to the PR, from an earlier check), might there be cases, even if in theory tests should have eliminated them, where a `{str}` key is discarding an int (say an ID value) that ought to be converted?

Reply:
If an expected `str` field is changed to an `int`, and the author ignored that `{str}` will filter, then yes. There are no realistic cases where that might happen intentionally however: Think about this: A field was previously a string, which holds arbitrary length characters. Now you get an integer, which is a number that holds less data (and of a different type). In what case would make a switch from the first to the second make sense, AND the result would make sense for a human? You cant generate more data from less data, the extractor has to be changed (perhaps the number is some index to get the actual data). Ergo filtering for `str` is generally the prefered way to not end up with potentially bogus data.

For numbers that game is flipped. You need a low data result and conversion from a string doesnt loose you that data (`"1234.5678"` => `1234.5678` is safe or close to safe)

Reply:
I was really thinking of site or hydration JSON where a field like `video_id` might be an integer after `json.loads()` is applied to the transferred JSON string, but actually IDs are one case that's almost sure to be caught in tests.